src/js/config.js

```diff
@@ -0,0 +1,127 @@
+/*
+Copyright 2015 OCAD University
+
+Licensed under the New BSD license. You may not use this file except in
+compliance with this License.
+
+You may obtain a copy of the License at
+https://raw.githubusercontent.com/GPII/first-discovery-server/master/LICENSE.txt
+*/
+
+"use strict";
+
+var fluid = require("infusion");
+var ajv = require("ajv");
+var $ = fluid.registerNamespace("jQuery");
+var gpii = fluid.registerNamespace("gpii");
+fluid.registerNamespace("gpii.firstDiscovery.server.config.oauth2.clientCredential");
+
+/**
+ * Attemps to load a config file, will return an empty object {} if the file
+ * cannot be located.
+ *
+ * @param file {String} - path to the config file
+ * @return {Object} - the JSON object representation of the JSON file, default is {}
+ */
+gpii.firstDiscovery.server.config.getConfigFile = function (file) {
+    var config = {};
+    try {
+        config = require(file);
+    } catch (e) {
+        console.log(file + " not found.");
+    }
+    return config;
+};
+
+/**
+ * Transforms an ENVMap into a config object
+ *
+ * @param ENVMap {Object} - a map of Environment variables that is transformed into a
+ *                          configuration object.
+ * @returns {Object} - the configuration object
+ */
+gpii.firstDiscovery.server.config.getENV = function (ENVMap) {
+    return fluid.transform(ENVMap, function (ENV) {
+        if (fluid.isPlainObject(ENV)) {
+            return gpii.firstDiscovery.server.config.getENV(ENV);
+        } else {
+            return process.env[ENV];
+        }
+    });
+};
+
+/**
+ * Returns the configuration object sourced from a configuration file and/or a
+ * set of environment variables. The environment variables take highest precedence
+ * and will override any value specified in the configuration file. Providing a
+ * schema will validate the resulting configuration object and will throw an error
+ * if the configuraiton is invalid. This is useful to ensure that the values are
+ * of the correct type and that all required properties are provided.
+ *
+ * @param file {String} - path to the config file
+ * @param ENVMap {Object} - a map of Environment variables that is transformed into a
+ *                          configuration object.
+ * @param schema {Object} - a JSON schema document based on AJV (see: https://github.com/epoberezkin/ajv)
+ * @returns {Object} - the combined and validated configuraiton object
+ */
+gpii.firstDiscovery.server.config.getConfig = function (file, ENVMap, schema) {
```

Comment:
This scheme isn't very extensible - and will make it increasingly hard to aggregate multiple sources of integration within config files in a declarative way. It suffers from the same problems of "function-ism" that we identified before in our prefs server driver - as soon as a configuration point consists of a function rather than a grade, we create a point of fragility which keeps propagating.
What I suggest instead is that you allow the user to just write standard declarative configs, with interpolation possible from the environment using a syntax such as "${environmentVariables}.GPII_OAUTH2_AUTH_GRANT_TYPE" etc.

A suggested scheme for this is to manually create a global component on startup which iterates over process.env and assigns each of the values found there into a top-level member.

I believe you're using a version of Infusion new enough to use the new "root" grade "fluid.resolveRootSingle" - http://docs.fluidproject.org/infusion/development/Contexts.html#global-components-fluid.resolveRoot-and-fluid.resolveRootSingle

By using this API you can ensure that the component you create, say, "gpii.environmentVariables" is resolvable by the above name from all configs.